### PR TITLE
6 develop visualrep

### PR DESCRIPTION
**Pull request by pontuslaestadius**

### CODE REVIEW

**Review by kosara-g — APPROVED** [review]
Good job, guys! Looking good!

**Review by ghost — APPROVED** [review]
Nice!

File: Application/Application.iml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (timjon) [inline Application/Application.iml:1]: Is this file needed?

File: Application/src/Main.java [hunk 2]
```diff
@@ -41,65 +48,72 @@ public void handle(ActionEvent event) {
...
+                    public void changed(ObservableValue<? extends Tab> ov, Tab t, Tab t1) {
+                        for (DiagramView dv: DiagramView.list) {
+                            dv.redraw();
+                        }
+                    }
+                }
+        );
 
-        stack.getChildren().add(pane);
+        BorderPane borderpane = new BorderPane();
```
Comment (timjon) [inline Application/src/Main.java]: What does line 80 to 86 do

File: Application/src/Main.java [hunk 3]
```diff
@@ -41,65 +48,72 @@ public void handle(ActionEvent event) {
...
+        borderpane.setLeft(ta);
+        borderpane.setCenter(tabPane);
 
-        Scene main = new Scene(stack,1200, 840);
+        Scene main;
+        //Scales the application to the size of the window or displays it in a maximized view.
+        Dimension screenSize = Toolkit.getDefaultToolkit().getScreenSize();
+        double width = screenSize.getWidth();
+        double height = screenSize.getHeight();
+        if (width < 1400) { // full screen
```
Comment (timjon) [inline Application/src/Main.java]: Why 1400 here?

File: Application/src/Main.java [hunk 4]
```diff
@@ -41,65 +48,72 @@ public void handle(ActionEvent event) {
...
+            primaryStage.setMaximized(true);
+        } else { // windowed
+            main = new Scene(borderpane,1280, 720);
+        }
 
         primaryStage.setScene(main);
         primaryStage.show();
-    }
 
+        try {
```
Comment (timjon) [inline Application/src/Main.java]: Why is there a sleep here

File: Application/src/Main.java [hunk 5]
```diff
@@ -41,65 +48,72 @@ public void handle(ActionEvent event) {
...
-        gc.strokePolygon(new double[]{60, 90, 60, 90},
-                new double[]{210, 210, 240, 240}, 4);
-        gc.strokePolyline(new double[]{110, 140, 110, 140},
-                new double[]{210, 210, 240, 240}, 4);
+        Draw.temp_generate_diagram(); // TODO replace with actual parsing.
+        try {
+            Thread.sleep(200); 
+        } catch (InterruptedException e) {
+            System.err.println(e.toString());
+            System.out.println(e.toString());
```
Comment (timjon) [inline Application/src/Main.java]: Duplicate lines of code

File: Application/src/Main.java [hunk 6]
```diff
@@ -41,65 +48,72 @@ public void handle(ActionEvent event) {
...
-        gc.strokeArc(60, 160, 30, 30, 45, 240, ArcType.CHORD);
-        gc.strokeArc(110, 160, 30, 30, 45, 240, ArcType.ROUND);
-        gc.fillPolygon(new double[]{10, 40, 10, 40},
-                new double[]{210, 210, 240, 240}, 4);
-        gc.strokePolygon(new double[]{60, 90, 60, 90},
-                new double[]{210, 210, 240, 240}, 4);
-        gc.strokePolyline(new double[]{110, 140, 110, 140},
-                new double[]{210, 210, 240, 240}, 4);
+        Draw.temp_generate_diagram(); // TODO replace with actual parsing.
+        try {
```
Comment (timjon) [inline Application/src/Main.java]: Random sleep again?

File: Application/src/visuals/Class.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+import javafx.scene.image.Image;
+import javafx.scene.paint.Color;
+
+import java.util.Iterator;
+
+/**
+ * A class wrapper for coordinates with extra properties that implements Renderable.
+ * @author Pontus Laestadius
+ */
+public class Class implements Renderable {
```
Comment (timjon) [inline Application/src/visuals/Class.java:13]: Should this really have the name class?

File: Application/src/visuals/Class.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    private Coordinates coordinates;
+    private int size;
+    private String name;
+
+    private int aniindex = 0;
+    private Coordinates[] anicontent;
```
Comment (timjon) [inline Application/src/visuals/Class.java]: What is line 18 and 19 used for

File: Application/src/visuals/Class.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        this.coordinates = coordinates;
+        this.size = size;
+    }
+
+    public void update() {
+        aniindex = aniindex == anicontent.length-1 ? 0: aniindex+1;
+    }
+
+    public String format() {
+        return null;
```
Comment (timjon) [inline Application/src/visuals/Class.java]: ?????

File: Application/src/visuals/Class.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    private String name;
+
+    private int aniindex = 0;
+    private Coordinates[] anicontent;
+
+    static Image castle = new Image("resources/castle.png");
+    static Image platform = new Image("resources/platform.png");
+    static Image pillar = new Image("resources/pillar.png");
+
+    public Class(String name) {
```
Comment (timjon) [inline Application/src/visuals/Class.java]: What does this do

File: Application/src/visuals/Class.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        this.name = name;
+        anicontent = new Coordinates[1];
+        anicontent[0] = new Coordinates(0,0);
+    }
+
+    public int getSize() {
+        return size;
+    }
+
+    public void place(Coordinates coordinates, int size) {
```
Comment (timjon) [inline Application/src/visuals/Class.java]: What does this do?

File: Application/src/visuals/Class.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+     */
+    public void render(GraphicsContext gc) {
+        gc.setFill(Color.TRANSPARENT);
+
+        int x = this.coordinates.getX();
+        int y = this.coordinates.getY();
+
+        int lifeline_width = size/8;
+        int scale = (int) (pillar.getHeight()/lifeline_width);
+        for (int i = 0; i < 1000; i+=scale) {
```
Comment (timjon) [inline Application/src/visuals/Class.java]: Random for-loop?

File: Application/src/visuals/Class.java [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+            gc.drawImage(pillar,x +size/4 -lifeline_width/2,y + size/6 +i,lifeline_width,scale);
+        }
+
+        gc.drawImage(platform,x -size/8,y + size/2 -size/8, size/2 + size/4,size/6);
```
Comment (timjon) [inline Application/src/visuals/Class.java]: Math??

File: Application/src/visuals/DiagramView.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+    /**
+     * @param name the name to match.
+     * @return true if the provided name is the selected tab.
+     */
+    public static boolean inView(String name) {
+        return tabPane.getSelectionModel().getSelectedItem().getText().equals(name);
+    }
+
+    public DiagramView(String tabName) {
```
Comment (timjon) [inline Application/src/visuals/DiagramView.java]: What does this do?

File: Application/src/visuals/DiagramView.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+        this.draw = new Draw(tabName, (int)tabPane.getWidth(), (int)tabPane.getHeight());
+        this.tabName = tabName;
+        list.add(this);
+    }
+
+    public Tab getTab() {
```
Comment (timjon) [inline Application/src/visuals/DiagramView.java]: Why is the getter creating a new tab

File: Application/src/visuals/Draw.java [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+ * @version 0.65
+ * @author Pontus Laestadius, Sebastian Fransson
+ */
+
+public class Draw {
+
+    private Canvas canvas;
+    private ArrayList<Class> classes = new ArrayList<>();
+    private ArrayList<Message> messages = new ArrayList<>(); // Stores the messages between nodes.
+    private int offset;
```
Comment (timjon) [inline Application/src/visuals/Draw.java]: Offset

File: Application/src/visuals/Draw.java [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+        for (Renderable r: messages)
+            r.update();
+    }
+
+    /**
+     * Renders the message when trying to resize the application.
+     */
+    void renderMessage() {
+        if(this.messages.size() > 0) {
+            for (int i = 0; i < messages.size(); i++) {
```
Comment (timjon) [inline Application/src/visuals/Draw.java]: What does this loop do?

File: Application/src/visuals/Draw.java [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+
+    /**
+     * Updates the class to fit the resized window.
+     */
+    void renderClass() {
+        if (classes.size() == 0) return; // There are no items to render
+        int space = (getWidth())/this.classes.size(); // The amount of space each class can use.
+        int size = space/2; // The size of the objects is half of it's given space.
+        class_size = size/2;
+        for(int i = 0; i < classes.size(); i++) {
```
Comment (timjon) [inline Application/src/visuals/Draw.java]: What does this loop do?

**Review by timjon — CHANGES_REQUESTED** [review]
I'll do more comments later

File: Application/src/visuals/Class.java (hunk 9 above)
Comment (pontuslaestadius) [reply Application/src/visuals/Class.java]: Part of interface, currently not used.

File: Application/Application.iml (hunk 1 above)
Comment (pontuslaestadius) [reply Application/Application.iml:1]: I need it to compile with IntelliJ so I would suggest keeping it.

File: Application/src/visuals/Class.java (hunk 13 above)
Comment (pontuslaestadius) [reply Application/src/visuals/Class.java]: Indeed. It is math!

File: Application/src/visuals/DiagramView.java (hunk 15 above)
Comment (pontuslaestadius) [reply Application/src/visuals/DiagramView.java]: Good point

File: Application/src/visuals/handlers/Animation.java [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+    public Animation(Draw item) {
+        this.item = item;
+    }
+
+    @Override
+    public void run() {
+         loop();
+    }
+
+    private void loop() {
```
Comment (timjon) [inline Application/src/visuals/handlers/Animation.java:23]: What does this do?

File: Application/src/visuals/handlers/Resizer.java [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+
+    /**
+     *
+     */
+    public void run() {
+        if (th == null) // Only allow one instance of the program to run at a time
+            time();
+    }
+
+    private void time(){
```
Comment (timjon) [inline Application/src/visuals/handlers/Resizer.java]: What does this do

**Review by timjon — CHANGES_REQUESTED** [review]

File: Application/src/visuals/handlers/Animation.java (hunk 19 above)
Comment (pontuslaestadius) [reply Application/src/visuals/handlers/Animation.java:23]: It loops. Pretty self explanatory with the document header.

**Review by timjon — APPROVED** [review]
